### DISCUSSION TITLE
Position in the formula

### DISCUSSION THREAD

**Discussion by emphasis87** [Excel-DNA/ExcelDna#683] [Q&A] [answered]
Is there a way to determine at which position in the formula is the current function call?

```csharp
[ExcelFunction]
public static object Func(object item)
{
  // At which position in the formula is this?
  return i;
}
```

`=Func(Func(1))`

I want to know which of the two `Func` calls I am currently executing. Is it possible?

**Comment by govert** [comment, marked as answer]
No way that I know of.

**Reply by keithalewis** [reply]
It is possible, but not easy. See: https://github.com/xlladdins/xll24/blob/d5028f8eb4d3e3f0a24ca987dc1f4fabf9809902/handle.h#L173

**Reply by emphasis87** [reply]
I don't understand what does the code do much. Can I extract a caller from a reference or something and build a call tree?

**Reply by keithalewis** [reply]
You have to keep track of the function(s) being called.

**Reply by emphasis87** [reply]
What I am worried in such approach is behaviour when `IsThreadSafe=true` is used.
Let's say we have `Func(Func(A1),Func(A2))` where `Func` calls backend and is a stream and periodically updated using RTD.
* Can the order of the `Func(A1)` and `Func(A2)` calls be unstable? If multiple threads are used for each `Func`, then yes, if a single thread is used for the whole cell then no.

As I understand it, RTD topic is bound to a cell through xlfCaller, therefore updating any of the 3 `Func` observables would cause reevaluation of the cell.
* Can this happen while the previous evaluation of the cell is still going on?

**Reply by emphasis87** [reply]
After some more reading, I understand it so, that you call RTD excel function in your UDF (or directly) and excel then connects the RTD topic to the cell. After signaling the topic is updated, excel asynchronously triggers refresh on appropriate cells. Excel also uses RTD server to ConnectData and DisconnectData to inform the server which topics are in use.

Overall it looks like tracking at which position in the formula we are is only possible when running in a single thread (IsThreadSafe=false).

**Comment by keithalewis** [comment]
Functions are called by Excel. You have limited control over the order in which they are called.
I wrote https://github.com/xlladdins/xll/blob/481178bced1022f0b36b582388b79a6eabe1a11f/xll/macrofun/xll_utility.cpp#L6 to manually enforce the order in which functions are called using my Excel add-in library. I don't think you can use that as an argument to a function however.
What makes you believe RTD is bound to a cell through xlfCaller?

**Reply by emphasis87** [reply]
Honestly, I am unsure how RTD is connected to a specific cell, that was my assumption from browsing AsyncObservableImpl.cs a bit.

**Reply by keithalewis** [reply]
One way to track a cell is by using a defined name. https://xlladdins.github.io/Excel4Macros/define.name.html
If a user moves the cell it refers to Excel will update the definition.
